### PR TITLE
Try to normalize associated types before processing obligations

### LINKED ISSUE #90729: Problem with GAT and HRTB
I tried this code:

```rust
#![feature(generic_associated_types)]

use std::marker::PhantomData;

pub trait Type {
    type Ref<'a>;
}

pub trait AsBytes {
    fn as_bytes(&self) -> &[u8];
}

impl AsBytes for &str {
    fn as_bytes(&self) -> &[u8] {
        (*self).as_bytes()
    }
}

pub struct Utf8;

impl Type for Utf8 {
    type Ref<'a> = &'a str;
}

pub struct Bytes<T: Type> {
    _marker: PhantomData<T>,
}

impl<T: Type> Bytes<T>
where
    for<'a> T::Ref<'a>: AsBytes,
{
    pub fn new() -> Self {
        Self {
            _marker: PhantomData,
        }
    }
}

fn main() {
    let _b = Bytes::<Utf8>::new();
}
```

I expected to see this happen: *I had expected compilation to succeed*

Instead, this happened: *Compilation fails with the error message below.*

```
error[E0277]: the trait bound `for<'a> <Utf8 as Type>::Ref<'a>: AsBytes` is not satisfied
  --> src\main.rs:41:14
   |
41 |     let _b = Bytes::<Utf8>::new();
   |              ^^^^^^^^^^^^^^^^^^ the trait `for<'a> AsBytes` is not implemented for `<Utf8 as Type>::Ref<'a>`
   |
note: required by `Bytes::<T>::new`
  --> src\main.rs:33:5
   |
33 |     pub fn new() -> Self {
   |     ^^^^^^^^^^^^^^^^^^^^
help: consider further restricting the associated type
   |
40 | fn main() where for<'a> <Utf8 as Type>::Ref<'a>: AsBytes {
   |           ++++++++++++++++++++++++++++++++++++++++++++++
```

### Meta

`rustc --version --verbose`:
```
rustc 1.58.0-nightly (e269e6bf4 2021-10-26)
binary: rustc
commit-hash: e269e6bf47f40c9046cd44ab787881d700099252
commit-date: 2021-10-26
host: x86_64-pc-windows-msvc
release: 1.58.0-nightly
LLVM version: 13.0.0
```

### PR DESCRIPTION
**Pull request by jackh726**
Closes #90729

r? @nikomatsakis

### CODE REVIEW

**Comment by nikomatsakis** [comment]
@jackh726 I don't recall us talking about this PR, have I just forgotten? Did you ever look at the perf results?

**Comment by jackh726** [comment]
@nikomatsakis I mentioned this at the end of our meeting on Monday.

As far as perf, I haven't done anything more than just looking through the benchmarks. It looks like `wg-grammar` was hit "hard"  ~3%, whereas everything else is 1% or less. In theory, there might be an optimization that could be done (maybe first check if there are projection types in `obligation.predicate`?). But this does fix a number of things that I think, on balance, mean the perf hit is okay.

File: compiler/rustc_trait_selection/src/traits/fulfill.rs [hunk 1]
```diff
@@ -355,7 +357,20 @@ impl<'a, 'b, 'tcx> FulfillProcessor<'a, 'b, 'tcx> {
 
         let infcx = self.selcx.infcx();
 
-        let binder = obligation.predicate.kind();
+        let mut obligations = Vec::new();
+        let predicate = crate::traits::project::try_normalize_with_depth_to(
```
Comment (jackh726) [inline compiler/rustc_trait_selection/src/traits/fulfill.rs:361]: TODO: check if obligation has projections first

File: compiler/rustc_trait_selection/src/traits/fulfill.rs [hunk 2]
```diff
@@ -355,6 +357,21 @@ impl<'a, 'b, 'tcx> FulfillProcessor<'a, 'b, 'tcx> {
...
+            let mut obligations = Vec::new();
+            let predicate = crate::traits::project::try_normalize_with_depth_to(
+                self.selcx,
+                obligation.param_env,
+                obligation.cause.clone(),
+                obligation.recursion_depth + 1,
+                obligation.predicate,
+                &mut obligations,
+            );
+            if predicate != obligation.predicate {
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/fulfill.rs:370]: another optimization here might be to not `return` if `obligations.is_empty()`

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/fulfill.rs:370]: Yeah, but if that's the only obligation (and we have no nested obligations that we need to check due to the normalization), then we can just update `obligation.predicate = predicate` like the shallow-resolve code below.

**Comment by jackh726** [comment]
@craterbot check

**Comment by compiler-errors** [comment]
Following up on my promise to investigate why issue-91139 regresses due to this PR: https://hackmd.io/Fq2yTjPtSumZRiuSjyDD9g cc: @nikomatsakis 

tl;dr is that WF [skips registering](https://github.com/rust-lang/rust/blob/master/compiler/rustc_trait_selection/src/traits/wf.rs#L346) many obligations if they have escaping bound vars. But if we normalize the `WellFormed` predicate before we get to the WF predicate code (i.e. we normalize in fulfillment, like in this PR), then we register the bounds we get during projection, which basically amounts to `for<'a> T: 'a` in that example. Hence the new outlives error.

(side note: there is a lot of code in wf that skips registering obligations if they capture escaping vars. this is kinda sus, but I can see how until we get a better idea of how implied bounds work with GATs, we probably can't fix this fully...)

**Comment by nikomatsakis** [comment]
@compiler-errors yes, that all makes sense (and happy to discuss the WF situation, it's one of the things I'd like to resolve this year, but indeed there are some steps along the way)

**Comment by nikomatsakis** [comment]
So I did a bit of investigation here. 

I was looking into keccak-check, which claims a 3% perf regression. 

However, the major change in this PR is that during fulfillment it does additional normalization. But when I added a `println!` call to that code, I found it never triggered during keccak at all.

@jackh726 am I right that there is no other obvious source of slowdown here?

**Comment by compiler-errors** [comment]
Should we queue up another perf run just to be sure?

**Comment by jackh726** [comment]
Honestly, I think it's just inlining changes. @compiler-errors had a PR that fixed the perf regression, then rebased and the fix went away. This code is hot enough that we've seen it before.

**Comment by nikomatsakis** [comment]
Adding the perf-regression-triaged label:

This PR results in some regressions (<3%) while making various correctness (but not soundness) fixes. After some investigation, these regressions don't seem to be related to the code being introduced (which, at least in the case of keccak, never even executes). Hence I'm going to r+ it.

**Comment by jackh726** [comment]
Actually, let's just @bors r=nikomatsakis

If crater does happen to give anything interesting, we can handle as appropriate.

**Comment by jackh726** [comment]
Apparently issue-91139 passes under nll

@bors r=nikomatsakis
